File: ts-sdk/client/tests/helpers/orca.ts
import { SPLASH_POOL_TICK_SPACING, swapInstructions, WHIRLPOOLS_CONFIG_ADDRESS } from "@orca-so/whirlpools";
import {
  fetchPosition,
  fetchWhirlpool,
  getFeeTierAddress,
  getInitializeAdaptiveFeeTierInstruction,
  getInitializeConfigInstruction,
  getInitializeFeeTierInstruction,
  getInitializePoolV2Instruction,
  getInitializePoolWithAdaptiveFeeInstruction,
  getInitializeRewardV2Instruction,
  getOracleAddress,
  getSetRewardEmissionsV2Instruction,
  getTickArrayAddress,
  getTokenBadgeAddress,
  getUpdateFeesAndRewardsInstruction,
  getWhirlpoolAddress,
} from "@orca-so/whirlpools-client";
import { getInitializableTickIndex, getTickArrayStartTickIndex, tickIndexToSqrtPrice } from "@orca-so/whirlpools-core";
import {
  type Address,
  GetAccountInfoApi,
  GetEpochInfoApi,
  GetMinimumBalanceForRentExemptionApi,
  GetMultipleAccountsApi,
  type IInstruction,
  Rpc,
  TransactionSigner,
} from "@solana/kit";
import { fetchMint, getMintToInstruction } from "@solana-program/token-2022";

import { getNextKeypair } from "./keypair.ts";
import { rpc, sendTransaction, signer } from "./mockRpc.ts";

export async function setupWhirlpoolsConfigAndFeeTiers(): Promise<Address> {
  const keypair = getNextKeypair();
  const instructions: IInstruction[] = [];

  instructions.push(
    getInitializeConfigInstruction({
      config: keypair,
      funder: signer,
      feeAuthority: signer.address,
      collectProtocolFeesAuthority: signer.address,
      rewardEmissionsSuperAuthority: signer.address,
      defaultProtocolFeeRate: 100,
    }),
  );

  const defaultFeeTierPda = await getFeeTierAddress(keypair.address, 128);
  instructions.push(
    getInitializeFeeTierInstruction({
      config: keypair.address,
      feeTier: defaultFeeTierPda[0],
      funder: signer,
      feeAuthority: signer,
      tickSpacing: 128,
      defaultFeeRate: 1000,
    }),
  );

  const concentratedFeeTierPda = await getFeeTierAddress(keypair.address, 64);
  instructions.push(
    getInitializeFeeTierInstruction({
      config: keypair.address,
      feeTier: concentratedFeeTierPda[0],
      funder: signer,
      feeAuthority: signer,
      tickSpacing: 64,
      defaultFeeRate: 300,
    }),
  );

  const splashFeeTierPda = await getFeeTierAddress(keypair.address, SPLASH_POOL_TICK_SPACING);
  instructions.push(
    getInitializeFeeTierInstruction({
      config: keypair.address,
      feeTier: splashFeeTierPda[0],
      funder: signer,
      feeAuthority: signer,
      tickSpacing: SPLASH_POOL_TICK_SPACING,
      defaultFeeRate: 1000,
    }),
  );

  const tickSpacing = 64;
  const adaptiveFeeTierPda = await getFeeTierAddress(keypair.address, 1024 + tickSpacing);
  instructions.push(
    getInitializeAdaptiveFeeTierInstruction({
      adaptiveFeeTier: adaptiveFeeTierPda[0],
      defaultBaseFeeRate: 3000,
      filterPeriod: 30,
      decayPeriod: 600,
      reductionFactor: 500,
      adaptiveFeeControlFactor: 4_000,
      maxVolatilityAccumulator: 350_000,
      delegatedFeeAuthority: signer.address,
      feeTierIndex: 1024 + tickSpacing,
      initializePoolAuthority: signer.address,
      majorSwapThresholdTicks: tickSpacing / 2,
      tickGroupSize: tickSpacing / 2,
      whirlpoolsConfig: keypair.address,
      funder: signer,
      feeAuthority: signer,
      tickSpacing: tickSpacing,
    }),
  );

  await sendTransaction(instructions);
  return keypair.address;
}

export async function setupWhirlpool(
  tokenA: Address,
  tokenB: Address,
  tickSpacing: number,
  config: { initialSqrtPrice?: bigint; adaptiveFee?: boolean } = {},
): Promise<Address> {
  const feeTierIndex = config.adaptiveFee ? 1024 + tickSpacing : tickSpacing;
  const feeTierAddress = await getFeeTierAddress(WHIRLPOOLS_CONFIG_ADDRESS, feeTierIndex);
  const whirlpoolAddress = await getWhirlpoolAddress(WHIRLPOOLS_CONFIG_ADDRESS, tokenA, tokenB, feeTierIndex);
  const oracleAddress = await getOracleAddress(whirlpoolAddress[0]);
  const vaultA = getNextKeypair();
  const vaultB = getNextKeypair();
  const badgeA = await getTokenBadgeAddress(WHIRLPOOLS_CONFIG_ADDRESS, tokenA);
  const badgeB = await getTokenBadgeAddress(WHIRLPOOLS_CONFIG_ADDRESS, tokenB);
  const mintA = await fetchMint(rpc, tokenA);
  const mintB = await fetchMint(rpc, tokenB);
  const programA = mintA.programAddress;
  const programB = mintB.programAddress;

  const sqrtPrice = config.initialSqrtPrice ?? tickIndexToSqrtPrice(0);

  const instructions: IInstruction[] = [];

  if (config.adaptiveFee) {
    instructions.push(
      getInitializePoolWithAdaptiveFeeInstruction({
        adaptiveFeeTier: feeTierAddress[0],
        initializePoolAuthority: signer,
        oracle: oracleAddress[0],
        tradeEnableTimestamp: null,
        whirlpool: whirlpoolAddress[0],
        tokenMintA: tokenA,
        tokenMintB: tokenB,
        whirlpoolsConfig: WHIRLPOOLS_CONFIG_ADDRESS,
        funder: signer,
        tokenVaultA: vaultA,
        tokenVaultB: vaultB,
        tokenBadgeA: badgeA[0],
        tokenBadgeB: badgeB[0],
        tokenProgramA: programA,
        tokenProgramB: programB,
        initialSqrtPrice: sqrtPrice,
      }),
    );
  } else {
    instructions.push(
      getInitializePoolV2Instruction({
        whirlpool: whirlpoolAddress[0],
        feeTier: feeTierAddress[0],
        tokenMintA: tokenA,
        tokenMintB: tokenB,
        tickSpacing,
        whirlpoolsConfig: WHIRLPOOLS_CONFIG_ADDRESS,
        funder: signer,
        tokenVaultA: vaultA,
        tokenVaultB: vaultB,
        tokenBadgeA: badgeA[0],
        tokenBadgeB: badgeB[0],
        tokenProgramA: programA,
        tokenProgramB: programB,
        initialSqrtPrice: sqrtPrice,
      }),
    );
  }

  await sendTransaction(instructions);
  return whirlpoolAddress[0];
}

export async function initializeReward(
  whirlpool: Address,
  rewardMintAddress: Address,
  rewardIndex: number,
  emissionsPerSecondX64: bigint,
) {
  const rewardMint = await fetchMint(rpc, rewardMintAddress);
  const rewardVault = getNextKeypair();
  const rewardTokenBadge = await getTokenBadgeAddress(WHIRLPOOLS_CONFIG_ADDRESS, rewardMintAddress);

  const instructions: IInstruction[] = [];
  instructions.push(
    getInitializeRewardV2Instruction({
      funder: signer,
      rewardAuthority: signer,
      rewardIndex,
      rewardMint: rewardMintAddress,
      rewardTokenBadge: rewardTokenBadge[0],
      rewardTokenProgram: rewardMint.programAddress,
      rewardVault,
      whirlpool,
    }),
  );

  instructions.push(
    getMintToInstruction(
      {
        mint: rewardMint.address,
        token: rewardVault.address,
        mintAuthority: signer,
        amount: (emissionsPerSecondX64 * 3600n * 24n) >> 64n,
      },
      { programAddress: rewardMint.programAddress },
    ),
  );

  instructions.push(
    getSetRewardEmissionsV2Instruction({
      emissionsPerSecondX64,
      rewardVault: rewardVault.address,
      rewardAuthority: signer,
      rewardIndex,
      whirlpool,
    }),
  );

  await sendTransaction(instructions);
}

export async function updateFeesAndRewards(positionAddress: Address) {
  const position = await fetchPosition(rpc, positionAddress);
  const whirlpoolAddress = position.data.whirlpool;
  const whirlpool = await fetchWhirlpool(rpc, whirlpoolAddress);

  const initializableLowerTickIndex = getInitializableTickIndex(
    position.data.tickLowerIndex,
    whirlpool.data.tickSpacing,
    false,
  );
  const initializableUpperTickIndex = getInitializableTickIndex(
    position.data.tickUpperIndex,
    whirlpool.data.tickSpacing,
    true,
  );

  const lowerTickArrayIndex = getTickArrayStartTickIndex(initializableLowerTickIndex, whirlpool.data.tickSpacing);
  const upperTickArrayIndex = getTickArrayStartTickIndex(initializableUpperTickIndex, whirlpool.data.tickSpacing);

  await sendTransaction([
    getUpdateFeesAndRewardsInstruction({
      position: positionAddress,
      tickArrayLower: (await getTickArrayAddress(whirlpoolAddress, lowerTickArrayIndex))[0],
      tickArrayUpper: (await getTickArrayAddress(whirlpoolAddress, upperTickArrayIndex))[0],
      whirlpool: whirlpoolAddress,
    }),
  ]);
}

export async function swapExactInputOrca(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi & GetMinimumBalanceForRentExemptionApi & GetEpochInfoApi>,
  signer: TransactionSigner,
  pool: Address,
  inputAmount: bigint,
  mint: Address,
  slippageToleranceBps?: number | undefined,
) {
  const { instructions } = await swapInstructions(
    rpc,
    {
      inputAmount,
      mint,
    },
    pool,
    slippageToleranceBps,
    signer,
  );
  await sendTransaction(instructions);
}

File: rust-sdk/client/src/tests/helpers/orca.rs
use crate::tests::RpcContext;
use orca_whirlpools::{swap_instructions, SwapType, SPLASH_POOL_TICK_SPACING, WHIRLPOOLS_CONFIG_ADDRESS};
use orca_whirlpools_client::{
    get_fee_tier_address, get_oracle_address, get_token_badge_address, get_whirlpool_address, InitializePoolV2, InitializePoolV2InstructionArgs,
    InitializePoolWithAdaptiveFee, InitializePoolWithAdaptiveFeeInstructionArgs, ADAPTIVE_FEE_TIER_DISCRIMINATOR, FEE_TIER_DISCRIMINATOR,
    WHIRLPOOLS_CONFIG_DISCRIMINATOR, WHIRLPOOL_ID,
};
use solana_program::{system_program, sysvar};
use solana_sdk::account::Account;
use solana_sdk::pubkey::Pubkey;
use solana_sdk::signature::{Signature, Signer};
use std::error::Error;

pub fn get_whirlpool_config_accounts(signer: &Pubkey) -> Vec<(Pubkey, Account)> {
    let mut accounts = vec![];

    let config = *WHIRLPOOLS_CONFIG_ADDRESS.lock().unwrap();
    accounts.push((
        config,
        Account {
            lamports: 100_000_000_000,
            data: [
                WHIRLPOOLS_CONFIG_DISCRIMINATOR,
                &signer.to_bytes(),
                &signer.to_bytes(),
                &signer.to_bytes(),
                &100u16.to_le_bytes(),
            ]
            .concat(),
            owner: WHIRLPOOL_ID,
            executable: false,
            rent_epoch: 0,
        },
    ));

    let default_fee_tier = get_fee_tier_address(&config, 128).unwrap().0;
    accounts.push((
        default_fee_tier,
        Account {
            lamports: 100_000_000_000,
            data: [FEE_TIER_DISCRIMINATOR, &config.to_bytes(), &128u16.to_le_bytes(), &1000u16.to_le_bytes()].concat(),
            owner: WHIRLPOOL_ID,
            executable: false,
            rent_epoch: 0,
        },
    ));

    let concentrated_fee_tier = get_fee_tier_address(&config, 64).unwrap().0;
    accounts.push((
        concentrated_fee_tier,
        Account {
            lamports: 100_000_000_000,
            data: [FEE_TIER_DISCRIMINATOR, &config.to_bytes(), &64u16.to_le_bytes(), &300u16.to_le_bytes()].concat(),
            owner: WHIRLPOOL_ID,
            executable: false,
            rent_epoch: 0,
        },
    ));

    let splash_fee_tier = get_fee_tier_address(&config, SPLASH_POOL_TICK_SPACING).unwrap().0;
    accounts.push((
        splash_fee_tier,
        Account {
            lamports: 100_000_000_000,
            data: [
                FEE_TIER_DISCRIMINATOR,
                &config.to_bytes(),
                &SPLASH_POOL_TICK_SPACING.to_le_bytes(),
                &1000u16.to_le_bytes(),
            ]
            .concat(),
            owner: WHIRLPOOL_ID,
            executable: false,
            rent_epoch: 0,
        },
    ));

    let tick_spacing = 64;
    let adaptive_fee_tier = get_fee_tier_address(&config, 1024 + tick_spacing).unwrap().0;
    accounts.push((
        adaptive_fee_tier,
        Account {
            lamports: 100_000_000_000,
            data: [
                ADAPTIVE_FEE_TIER_DISCRIMINATOR,
                &config.to_bytes(),
                &(1024_u16 + tick_spacing).to_le_bytes(), // fee_tier_index
                &tick_spacing.to_le_bytes(),              // tick_spacing
                &signer.to_bytes(),                       // initialize_pool_authority
                &signer.to_bytes(),                       // delegated_fee_authority
                &3000u16.to_le_bytes(),                   // default_base_fee_rate
                &30u16.to_le_bytes(),                     // filter_period
                &600u16.to_le_bytes(),                    // decay_period
                &500u16.to_le_bytes(),                    // reduction_factor
                &4000u32.to_le_bytes(),                   // adaptive_fee_control_factor
                &350000u32.to_le_bytes(),                 // max_volatility_accumulator
                &(tick_spacing / 2).to_le_bytes(),        // tick_group_size
                &(tick_spacing / 2).to_le_bytes(),        // major_swap_threshold_ticks
                &[0_u8; 128],
            ]
            .concat(),
            owner: WHIRLPOOL_ID,
            executable: false,
            rent_epoch: 0,
        },
    ));

    accounts
}

pub async fn setup_whirlpool(
    ctx: &RpcContext,
    token_a: &Pubkey,
    token_b: &Pubkey,
    tick_spacing: u16,
    initial_sqrt_price: u128,
    adaptive_fee: bool,
) -> Result<Pubkey, Box<dyn Error>> {
    let account_infos = ctx.rpc.get_multiple_accounts(&[*token_a, *token_b])?;
    let mint_a_info = account_infos[0].as_ref().ok_or(format!("Mint {} not found", token_a))?;
    let token_program_a = mint_a_info.owner;
    let mint_b_info = account_infos[1].as_ref().ok_or(format!("Mint {} not found", token_b))?;
    let token_program_b = mint_b_info.owner;

    let pool_address = get_whirlpool_address(&*WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?, &token_a, &token_b, tick_spacing)?.0;
    let fee_tier_index = if adaptive_fee { 1024 + tick_spacing } else { tick_spacing };
    let fee_tier = get_fee_tier_address(&*WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?, fee_tier_index)?.0;
    let token_badge_a = get_token_badge_address(&*WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?, &token_a)?.0;
    let token_badge_b = get_token_badge_address(&*WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?, &token_b)?.0;
    let oracle = get_oracle_address(&pool_address)?.0;

    let token_vault_a = ctx.get_next_keypair();
    let token_vault_b = ctx.get_next_keypair();

    let mut instructions = vec![];

    if adaptive_fee {
        instructions.push(
            InitializePoolWithAdaptiveFee {
                whirlpools_config: *WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?,
                token_mint_a: *token_a,
                token_mint_b: *token_b,
                token_badge_a,
                token_badge_b,
                funder: ctx.signer.pubkey(),
                initialize_pool_authority: ctx.signer.pubkey(),
                whirlpool: pool_address,
                adaptive_fee_tier: fee_tier,
                oracle,
                token_vault_a: token_vault_a.pubkey(),
                token_vault_b: token_vault_b.pubkey(),
                token_program_a,
                token_program_b,
                system_program: system_program::id(),
                rent: sysvar::rent::ID,
            }
            .instruction(InitializePoolWithAdaptiveFeeInstructionArgs {
                initial_sqrt_price,
                trade_enable_timestamp: None,
            }),
        );
    } else {
        instructions.push(
            InitializePoolV2 {
                whirlpools_config: *WHIRLPOOLS_CONFIG_ADDRESS.try_lock()?,
                token_mint_a: *token_a,
                token_mint_b: *token_b,
                token_badge_a,
                token_badge_b,
                funder: ctx.signer.pubkey(),
                whirlpool: pool_address,
                token_vault_a: token_vault_a.pubkey(),
                token_vault_b: token_vault_b.pubkey(),
                fee_tier,
                token_program_a,
                token_program_b,
                system_program: system_program::id(),
                rent: sysvar::rent::ID,
            }
            .instruction(InitializePoolV2InstructionArgs {
                initial_sqrt_price,
                tick_spacing,
            }),
        );
    }

    ctx.send_transaction_with_signers(instructions, vec![&token_vault_a, &token_vault_b])?;

    Ok(pool_address)
}

pub async fn swap_exact_in(
    ctx: &RpcContext,
    whirlpool_address: &Pubkey,
    input_amount: u64,
    mint_address: &Pubkey,
    slippage_tolerance_bps: Option<u16>,
) -> Result<Signature, Box<dyn Error>> {
    ctx.send_transaction(
        swap_instructions(
            ctx.rpc.get_inner_client(),
            *whirlpool_address,
            input_amount,
            *mint_address,
            SwapType::ExactIn,
            slippage_tolerance_bps,
            Some(ctx.signer.pubkey()),
        )
        .await
        .unwrap()
        .instructions,
    )
}
